core/renderers/measurables/rows.js

```diff
@@ -0,0 +1,389 @@
+goog.require('Blockly.blockRendering.Measurable');
+goog.require('Blockly.blockRendering.NextConnection');
+goog.require('Blockly.blockRendering.PreviousConnection');
+goog.require('Blockly.RenderedConnection');
+
+/**
+ * An object representing a single row on a rendered block and all of its
+ * subcomponents.
+ * @package
+ * @constructor
+ */
+Blockly.blockRendering.Row = function() {
+  /**
+   * The type of this rendering object.
+   * @package
+   * @type {string}
+   */
+  this.type = 'row';
+
+  /**
+   * An array of elements contained in this row.
+   * @package
+   * @type {!Array.<!Blockly.blockRendering.Measurable>}
+   */
+  this.elements = [];
+
+  // TODO: Should row extend measurable?  Some of these properties are shared.
+  /**
+   * The height of the row.
+   * @package
+   * @type {number}
+   */
+  this.height = 0;
+
+  /**
+   * The width of the row, from the left edge of the block to the right.
+   * Does not include child blocks unless they are inline.
+   * @package
+   * @type {number}
+   */
+  this.width = 0;
+
+  /**
+   * The width of the row, from the left edge of the block to the edge of the
+   * block or any connected child blocks.
+   * @package
+   * @type {number}
+   */
+  this.widthWithConnectedBlocks = 0;
+
+  /**
+   * The Y position of the row relative to the origin of the block's svg group.
+   * @package
+   * @type {number}
+   */
+  this.yPos = 0;
+
+  /**
+   * The X position of the row relative to the origin of the block's svg group.
+   * @package
+   * @type {number}
+   */
+  this.xPos = 0;
+
+  /**
+   * Whether the row has any external inputs.
+   * @package
+   * @type {boolean}
+   */
+  this.hasExternalInput = false;
+
+  /**
+   * Whether the row has any statement inputs.
+   * @package
+   * @type {boolean}
+   */
+  this.hasStatement = false;
+
+  /**
+   * Whether the row has any inline inputs.
+   * @package
+   * @type {boolean}
+   */
+  this.hasInlineInput = false;
+
+  /**
+   * Whether the row has any dummy inputs.
+   * @package
+   * @type {boolean}
+   */
+  this.hasDummyInput = false;
+
+  /**
+   * Whether the row has a jagged edge.
+   * @package
+   * @type {boolean}
+   */
+  this.hasJaggedEdge = false;
+};
+
+/**
+ * The shape object to use when drawing previous and next connections.
+ * TODO (#2803): Formalize type annotations for these objects.
+ * @type {Object}
+ */
+Blockly.blockRendering.Row.prototype.notchShape =
+    Blockly.blockRendering.constants.NOTCH;
+
+/**
+ * Whether this is a spacer row.
+ * @return {boolean} Always false.
+ * @package
+ */
+Blockly.blockRendering.Row.prototype.isSpacer = function() {
+  return false;
+};
+
+/**
+ * Inspect all subcomponents and populate all size properties on the row.
+ * @package
+ */
+Blockly.blockRendering.Row.prototype.measure = function() {
+  var connectedBlockWidths = 0;
+  for (var e = 0; e < this.elements.length; e++) {
+    var elem = this.elements[e];
+    this.width += elem.width;
```

Comment:
This is a preexisting issue, but since this.width is not reset, I think the width is incorrect if the method is ever accidentally called twice. Potentially we could introduce a minWidth property that this.width is reset to in the beginning.